### ISSUE TITLE
Texture Mapping: Missing Rotation

### ISSUE THREAD

**Issue by Ekki** [mrdoob/three.js#5808]
Hi,

in the meta model, i.e. the data model from which I create my ThreeJS scenes, I'm using the X3D texture mapping manipulation as defined here: http://doc.x3dom.org/author/Texturing/TextureTransform.html

while scaling and offset seem to be there, rotation is missing. Is there any chance to get the rotation in the future?

Thx and BR.

   Ekki

**Comment by Ekki** [comment]
Happy New Year guys, BTW...

Again, Ben, this looks great! Any time schedule when this could be available to the public? Is there anything I can do to support you? BR, Ekki

**Comment by bhouston** [comment]
@Ekki I think we need a decision from @mrdoob and @WestLangley as to whether they want to add this to the shaders or as utility to transform UVs on meshes prior to uploading to the card.  It is their choice.

**Comment by bhouston** [comment]
@Ekki This can be implemented as a loader functionality outside of the core ThreeJS as long as the UVs are not animated.  In that case it is really simple to do.

**Comment by mrdoob** [comment]
I think the `Matrix3` approach is the way to go. Messing with UVs is... a mess.

**Comment by mrdoob** [comment]
We could also, potentially, have a different `Matrix3` per texture.

**Comment by bhouston** [comment]
If we go the Matrix3 route, we should have a different Matrix3 per texture. :)  I have some plans to redo the texture system so that you can map any texture to any UV, to any input (map, emissiveMap, alphaMap, etc.)  But that will come later.

**Comment by Ekki** [comment]
I would really like that Matrix3-per-Texture way! So I could share the geometries globally and assign 'mapped' textures to each instance. Sounds like a deal! ;-)

**Comment by bhouston** [comment]
Myself, I am going to be a little gunshy about writing new PRs until I know the fate of the existing PRs as I am worried about conflicts and whether or not I can make use of common.glsl, and diffuseColor/outgoingLight in new PRs.

**Comment by WestLangley** [comment]
I do not think that users should be expected to specify a matrix. You can implement it that way, but there should be a more natural API -- something like translate, rotate, scale, center.

Rotation is naturally around the center, but scaling can either be from the corner or the center, so I am not sure how that should work.

>  I am going to be a little gun-shy about writing new PRs

That is completely understandable. I also would like the PRs merged one-at-a-time so I have an opportunity to review them as they are applied.

**Comment by bhouston** [comment]
@WestLangley I agree we shouldn't use a Matrix3 as the interface.  BTW here is the code I wrote for Clara.io to transform the UVs (influenced by our conversations on this issue):

```
  getUVTransform: function( rotation, tiling, useScale, scale, translation, pivotCenter ) {

    var result = new THREE.Matrix4();

    result.multiply( new THREE.Matrix4().makeTranslation( -pivotCenter.x, -pivotCenter.y, 0 ) );
    result.multiply( new THREE.Matrix4().makeRotationZ( THREE.Math.degToRad( rotation ) ) );
    if( useScale ) {
      result.multiply( new THREE.Matrix4().makeScale( scale.x, scale.y, 0 ) );
    }
    else {
      var safeInverse = function( v ) {
        if( v == 0 ) {
          return 1;
        }
        return 1 / v;
      }
      result.multiply( new THREE.Matrix4().makeScale( safeInverse( tiling.x ), safeInverse( tiling.y ), 0 ) );      
    }
    result.multiply( new THREE.Matrix4().makeTranslation( pivotCenter.x, pivotCenter.y, 0 ) );
    result.multiply( new THREE.Matrix4().makeTranslation( translation.x, translation.y, 0 ) );

    return result;

  },

var applyUVTransform = function() {
  var tempVec3 = new THREE.Vector3();
  return function( uv, matrix ) {
    tempVec3.set( uv.x, uv.y, 0 );
    tempVec3.applyMatrix4( matrix );
    uv.set( tempVec3.x, tempVec3.y );      
  }
}();
```

I used a Matrix4 because Matrix3 didn't have the necessary tools (makeRotation, makeTranslation, makeScale.)

**Comment by WestLangley** [comment]
@bhouston I was under the assumption that we would construct the matrix on the javascript side from the user's natural parameters. Something like

    matrix3 = new THREE.Matrix3().makeUVTransform( rotate, scale, translate, optionalCenter );

Then, instead of passing `offsetRepeat` to the shader, we would pass `matrix3` instead.

Is that your understanding?

**Comment by bhouston** [comment]
@WestLangley I agree.  The code above can be used, to convert from the Matrix4 I construct to Matrix3 for shader usage just drop column 3 and row 3.

**Comment by mrdoob** [comment]
> Myself, I am going to be a little gunshy about writing new PRs until I know the fate of the existing PRs as I am worried about conflicts and whether or not I can make use of common.glsl, and diffuseColor/outgoingLight in new PRs.

Sorry for the delay. I'm holding these until the 15th. There was a release blocker with morphtargets for the last weeks. Finally figured it out this week. Now I just want to have scripting in the editor in a useful state and will be ready for release on the 15th. Until then I'll be merging only PRs that are simple.

**Comment by kkananen** [comment]
@mrdoob any update on this? We at vizor.io are implementing this and don't want to diverge too far from your plans. Any new info much appreciated.

**Comment by mrdoob** [comment]
I'm happy to merge whatever @bhouston and @WestLangley decide.

**Comment by Kesshi** [comment]
@kkananen Did you finish your implementation of this feature?
Maybe you could make a PR?
For my current project the missing rotation is also big issue.

**Comment by nxgraphics** [comment]
any status on this ? is it implemented ? is there an example somewhere ? thx

**Comment by Mugen87** [comment]
Solved with #11863

https://threejs.org/examples/webgl_materials_texture_rotation.html
